xtm/xtm_api.cc

```diff
@@ -0,0 +1,353 @@
+	} else {
+		queue->producer_read_fd = -1;
+		queue->producer_write_fd = -1;
+	}
+	queue->is_producer_need_notifications = is_producer_need_notifications;
+	queue->is_need_to_close_readfd = is_need_to_close_readfd;
+	queue->is_notification_expected = false;
+
+	if (fcntl(queue->consumer_read_fd, F_SETFL, O_NONBLOCK) < 0 ||
+	    fcntl(queue->consumer_write_fd, F_SETFL, O_NONBLOCK) < 0 ||
+	    (is_producer_need_notifications &&
+	    (fcntl(queue->producer_read_fd, F_SETFL, O_NONBLOCK) < 0 ||
+	     fcntl(queue->producer_write_fd, F_SETFL, O_NONBLOCK) < 0))) {
+		save_errno = errno;
+		goto close_producer_fd;
+	}
+	if (queue->queue.create(size) < 0) {
+		save_errno = EINVAL;
+		goto close_producer_fd;
+	}
+	return queue;
+
+close_producer_fd:
+	if (is_producer_need_notifications) {
+		close(queue->producer_read_fd);
+		if (queue->producer_read_fd != queue->producer_write_fd)
+			close(queue->producer_write_fd);
+	}
+close_consumer_fd:
+	close(queue->consumer_read_fd);
+	if (queue->consumer_read_fd != queue->consumer_write_fd)
+		close(queue->consumer_write_fd);
+free_queue:
+	free(queue);
+	errno = save_errno;
+	return NULL;
+};
+
+int
+xtm_queue_delete(struct xtm_queue *queue)
+{
+	int rc = 0;
+	if (queue->is_need_to_close_readfd &&
+	    queue->is_producer_need_notifications &&
+	    close(queue->producer_read_fd) < 0)
+		rc = -1;
+	if (queue->is_producer_need_notifications &&
+	    queue->producer_write_fd != queue->producer_read_fd &&
+	    close(queue->producer_write_fd) < 0)
+		rc = -1;
+	if (queue->is_need_to_close_readfd &&
+	    close(queue->consumer_read_fd) < 0)
+		rc = -1;
+	if (queue->consumer_read_fd != queue->consumer_write_fd &&
+	    close(queue->consumer_write_fd) < 0)
+		rc = -1;
+	free(queue);
+	return rc;
+}
+
+int
+xtm_queue_notify(struct xtm_queue *queue)
+{
+	return notify_fd(queue->consumer_write_fd);
+}
+
+int
+xtm_queue_probe(struct xtm_queue *queue)
+{
+	if (queue->queue.free_count() == 0) {
+		errno = ENOBUFS;
+		return -1;
+	}
+	return 0;
+}
+
+unsigned
+xtm_queue_count(struct xtm_queue *queue)
+{
+	return queue->queue.count();
+}
+
+int
+xtm_queue_push_fun(struct xtm_queue *queue, xtm_queue_fun_t fun,
+		   void *fun_arg)
+{
+	union xtm_msg xtm_msg;
+	xtm_msg.fun = fun;
+	xtm_msg.fun_arg = fun_arg;
+	if (queue->queue.put(&xtm_msg, 1) != 1) {
+		errno = ENOBUFS;
+		if (queue->is_producer_need_notifications) {
+			__atomic_store_n(&queue->is_notification_expected,
+					 true, __ATOMIC_RELEASE);
+		}
+		return -1;
+	}
+	return 0;
+}
+
+int
+xtm_queue_consumer_fd(struct xtm_queue *queue)
+{
+	return queue->consumer_read_fd;
+}
+
+int
+xtm_queue_producer_fd(struct xtm_queue *queue)
+{
+	return queue->producer_read_fd;
+}
+
+int
+xtm_queue_invoke_funs_all(struct xtm_queue *queue)
+{
+	struct xtm_scsp_queue_read_iterator<xtm_msg> iter;
+	const union xtm_msg *xtm_msg;
+	unsigned cnt = 0;
+
+	iter.begin(&queue->queue);
+	while((xtm_msg = iter.read()) != nullptr) {
+		xtm_msg->fun(xtm_msg->fun_arg);
+		cnt++;
+	}
```

Comment:
It may (or may not) be a good idea to check is_notification_expected on every iteration. Otherwise writer can be stuck for a LONG time on idle CPU until reader calls everything.

Reply:
Doing so would also work around race between reading and writing to is_notification_expected on different CPUs - there is a short time interval between producer decision to fail and writing to is_notification_expected, reading of is_notification_expected by consumer can happen in that interval.

Reply:
To prevent situations when consumer and producer threads sleep, waiting inifinity for each other you need to call xtm_queue_notify in case when xtmp_queue_push* fails! See perf/xtm.cc for example! If you do it, race
between reading and writing to is_notification_expected  is not problem. But may be you idea to check this flag is not bad idea, because you right, producer can be stuck for a long time

Reply:
@Gerold103 What do you think about it?

Reply:
Talking of the issue about having to call notify in case of push fail. Here is the schema you have now:
```
producer:
    - fail to put  (1)
    - store        (2)
    - wait

consumer:
    - pop          (3)
    - load         (4)
    - clear        (5)
    - notify       (6)
```
The problem you have: after step 1 the steps 3-4 might happen. Steps 5-6 will not happen ever because on the step 4 you saw the flag was false - nobody was waiting. But you can fix it like this:


```
producer:
    - fail to put      (1)
    - store            (2)
    - try to put again (3)
      if failed, then return -1

consumer:
    - pop          (4)
    - load         (5)
    - clear        (6)
    - notify       (7)
```
Even if after step 1 the steps 4-5 are done in another thread, you can be calm. On the step 3 you will see the queue is not full again. Because step 4 was done. If it is not done yet, then you managed to do the store in 2, and load in 5 will see it and do the notify.

No need to check anything in a loop. Will it work? I suggest to try it in your bench + remove the notification in case of push fail. It should work smooth.

Reply:
> To prevent situations when consumer and producer threads sleep, waiting inifinity for each other you need to call xtm_queue_notify in case when xtmp_queue_push* fails!

Then why doesn't xtm_queue_push*() call xtm_queue_notify()? Why you are relying on careless programmers which do not read docs (you haven't documented such requirement BTW).

Reply:
> And the main issue - if you allow to push more while you read,

Of course, it is allowed - this happens on another CPU. E. g. h2o threads feed TX thread with HTTP requests.

> you are guaranteed to have starvation. Because this loop will run forever if the producer is fast enough.

No, it will not - queue->read is not updated inside execution loop so producer will see "no space".

Reply:
>Then why doesn't xtm_queue_push*() call xtm_queue_notify()? Why you are relying on careless programmers which do not read docs (you haven't documented such requirement BTW).

It was decided we don't want any implicit notifications to happen AFAIR.

Reply:
>No, it will not - queue->read is not updated inside execution loop so producer will see "no space".

Exactly, because read is not updated. You propose to notify the producer in the loop, but it makes zero sense, because it won't be able to push anyway. But if you update the read value, you have the infinite loop and starvation.

Reply:
>But there is a little problem in this case consumer thread very very rare notifies producer thread, when it's really not needed.

It is fine as long as it happens very rarely and only when the queue gets full. We do not optimize the full queue case as it is not supposed to happen really. It should work, but not necessarily in a most efficient way.

Reply:
> > No, it will not - queue->read is not updated inside execution loop so producer will see "no space".
> 
> Exactly, because read is not updated. You propose to notify the producer in the loop, but it makes zero sense, because it won't be able to push anyway.

Of course read must be updated. You mentioned that yourself: "Here the problem is not only in the checking the flag - you would also need to commit the read iterator and open it again to continue."

> But if you update the read value, you have the infinite loop and starvation.

I have no problem with infinite loop, in httpng that's separate fiber which creates and starts a lot of per-request fibers which yield. Implementing upper limit on calls count may be a good idea - but there should be efficient and simple way to disable it (not with INT_MAX etc - it WOULD overflow).

Reply:
> Of course read must be updated. You mentioned that yourself

I mentioned it specifically as a bad idea. Which shall not be done ever. Otherwise TX thread is going to stall in the loop of invoking more and more functions.

> I have no problem with infinite loop, in httpng that's separate fiber which creates and starts a lot of per-request fibers which yield

Yeah, I see now. I looked at httpng sources. There are problems though still.
* xtm is not a part of httpng. It is a separate module and it can not hope that the invoked functions would yield. And it will not work if the functions do not yield. The loop might simply run infinitely not letting the other fibers work, or just run too long, which is not acceptable either;
* Not all of your functions yield AFAIS. So you still might enter a too long invoke strike without yields;

That is not how cooperative multistasking should work. If you would allow to repopulate the queue while it is being read, there would not be any protection against the starvation of other fibers. You need to do a part of work, not more than a certain limit, and then switch to other work (yield) to continue on the next iteration of the event loop.

Your current implementation, btw, from what I see, might be a bit of an overkill. You could use `ev_io` and attach your descriptor directly to the loop. Then you would get notifications without having to start a fiber. But that is out of scope of this PR.

>  Implementing upper limit on calls count may be a good idea - but there should be efficient and simple way to disable it 

It is disabled. Because `invoke_funcs_all()` shall not allow to repopulate the queue while it works. It is simply against the cooperative multitasking as I tried to explain above. I was more thinking about adding a new function `invoke_funcs(count)`, without `_all` suffix. It would take max number of functions to invoke. It might be useful if the queue size is millions and you obviously can't afford calling so many virtual functions in a single batch, blocking the scheduler.